law/task/base.py

```diff
@@ -230,14 +231,19 @@ def __init__(self, *args, **kwargs):
         self._task_logger = None
 
     def complete(self):
+        # create a flat list of all non-optional outputs
+        # target collections handle optional outputs themselves in their exists() call
         outputs = [t for t in flatten(self.output()) if not t.optional]
 
         if len(outputs) == 0:
             logger.warning("task {!r} has either no non-optional outputs or no custom complete() "
                 "method".format(self))
-            return False
+            return True
 
-        return all(t.exists() for t in outputs)
+        return all(
+            t.exists(skip_optional=True) if isinstance(t, TargetCollection) else t.exists()
+            for t in outputs
+        )
```

Comment:
Regarding your [comment above](https://github.com/lmoureaux/law/pull/1#pullrequestreview-1302071088), `skip_optional` is set to `True` here.

Reply:
Hmm, from a maintainability perspective I don't like how support for this invades all parts of Law (with `isinstance` instead of proper OOP)

Reply:
Perfectly true. Wasn’t too happy with that either, so I did a refactoring now that I‘ll push later ;)